nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/smtp/handler/SMTPMessageHandlerFactory.java

```diff
@@ -0,0 +1,147 @@
+        public void recipient(String recipient) throws RejectException {
+            // TODO: possibly whitelist receivers?
+            this.recipient = recipient;
+        }
+
+        @Override
+        public void data(InputStream inputStream) throws RejectException, TooMuchDataException, IOException {
+            // Start counting the timer...
+
+            StopWatch watch = new StopWatch(false);
+
+            SMTPServer server = messageContext.getSMTPServer();
+
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+            byte [] buffer = new byte[1024];
+            int rd;
+
+            while ((rd = inputStream.read(buffer, 0, buffer.length)) != -1) {
+                baos.write(buffer, 0, rd);
+            }
+            if (baos.getBufferLength() > server.getMaxMessageSize()) {
+                throw new TooMuchDataException("Data exceeds the amount allowed.");
+            }
+
+            baos.flush();
+            this.messageBody = baos.toByteArray();
+
+
+            X509Certificate[] certificates = new X509Certificate[]{};
+
+            String remoteIP = messageContext.getRemoteAddress().toString();
+            String helo = messageContext.getHelo();
+            if (messageContext.getTlsPeerCertificates() != null ){
+                certificates = (X509Certificate[]) messageContext.getTlsPeerCertificates().clone();
+            }
+
+            SmtpEvent message = new SmtpEvent(remoteIP, helo, from, recipient, certificates, messageBody);
+            try {
+                    // Try to queue the message back to the NiFi session
+                    incomingMessages.put(message);
+            } catch (InterruptedException e) {
+                // Throws an error to the  incoming server alerting about issues queuing the message in NiFi
+                logger.error("Hit an error sending message back to NiFi main thread {}. Sending and error back to SMTP client and discarding message", new Object [] {e});
+                throw new DropConnectionException(451, "NiFi something went wrong while processing you message, please retry again later");
+            }
+
+            // Once message has been sent to the queue, it should be processed by NiFi onTrigger,
+            // a flowfile created and its processed status updated before an acknowledgment is
+            // given back to the SMTP client
+            try {
+                synchronized(message) {
+                    while(! message.getProcessed()) {
+                        // Check to see if it is too late...
+                        final long serverTimeout = TimeUnit.NANOSECONDS.convert(messageContext.getSMTPServer().getConnectionTimeout(), TimeUnit.MILLISECONDS);
+                        if ( watch.getElapsed(TimeUnit.MILLISECONDS) <=  serverTimeout) {
+                            // Will should allow for a couple of retries;
+                            message.wait(messageContext.getSMTPServer().getConnectionTimeout() / 20);
```

Comment:
I think it's Java related issue, so it can't be reproduced by telnet command.

The error is reproducible in my environment using PutEmail and ListenSMTP with `Maximum number of SMTP connection` as 1, without SSL configuration. The reason is PutEmail reconnects by its fallback mechanism. Connection requests were sent twice before ListenSMTP closes the first one. Please take a look on another comment I sent.